### PR TITLE
cython modifications

### PR DESCRIPTION
**Pull request by AshKelly**
Hi all,

I just thought I'd share some of the code changes I have made and the impact they've had.

# TLDR
I have managed to achieve about a **factor of 2** speed up on the simple example Jacob shared with me, `test_arcticpy/profile_demo.py`

On my local machine this problem outputs:

```
express = 5
output_name = "test_5"
do_plot = False
2067 row(s), 10 column(s)

         21243410 function calls (19635936 primitive calls) in 32.943 seconds
```

as opposed to `28045046 function calls (26951282 primitive calls) in 61.871 seconds` on the `dev` branch. I've also added a local (uncommitted) check which ensures the output remains identical.

# What did I change?

The largest performance increase actually came from partial cythonizing of `clock_charge_in_one_direction`. A simple line profiler suggested that this function was wasting a lot of time doing simple loops. 

I also got performance gains from the following methods
`n_trapped_electrons_from_watermarks`
`update_watermark_volumes_for_cloud_below_highest` (particularly removing the `np.roll`)
and cythonizing a small part of `watermark_index_above_cloud_from_cloud_fractional_volume`

I also found some code like the following, 
`if a not in np.cumsum(an_array)`

This particularly line was wasting quite a lot of time. I wrote a small cython function which achieves the same result and can be used as follows,

`if not  cy_value_in_cumsum(a, an_array)`

This is entirely C, and doesn't require any array allocations.

### CODE REVIEW

**Comment by AshKelly** [comment]
I'm not entirely sure where else there are more gains to be made. The profiler outputs:

```
   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   33.429   33.429 test_arcticpy/profile_demo.py:82(add_cti_to_hst_image)
        1    0.000    0.000   33.315   33.315 /home/ash/Projects/arcticpy/arcticpy/main.py:269(add_cti)
        1    2.481    2.481   33.315   33.315 arcticpy/main_utils.pyx:15(cy_clock_charge_in_one_direction)
    82670    2.524    0.000   27.320    0.000 /home/ash/Projects/arcticpy/arcticpy/trap_managers.py:700(n_electrons_released_and_captured)
   123996    1.354    0.000    8.115    0.000 /home/ash/Projects/arcticpy/arcticpy/trap_managers.py:494(collapse_redundant_watermarks)
    61255    1.520    0.000    7.183    0.000 /home/ash/Projects/arcticpy/arcticpy/trap_managers.py:619(_n_electrons_released_and_captured__cloud_below_watermarks)
641386/517396    0.609    0.000    5.706    0.000 {built-in method numpy.core._multiarray_umath.implement_array_function}
   247275    0.654    0.000    4.434    0.000 /home/ash/Projects/arcticpy/arcticpy/trap_managers.py:351(n_trapped_electrons_from_watermarks)
    82670    2.315    0.000    3.555    0.000 /home/ash/Projects/arcticpy/arcticpy/trap_managers.py:302(fill_probabilities_from_dwell_time)
   123990    0.097    0.000    3.423    0.000 <__array_function__ internals>:2(roll)
   123990    2.301    0.000    3.055    0.000 /home/ash/Projects/arcticpy/venv/lib/python3.8/site-packages/numpy-1.19.2-py3.8-linux-x86_64.egg/numpy/core/numeric.py:1113(roll)
   243712    0.251    0.000    2.399    0.000 /home/ash/Projects/arcticpy/arcticpy/trap_managers.py:405(update_watermark_volumes_for_cloud_below_highest)
    20670    0.052    0.000    2.210    0.000 /home/ash/Projects/arcticpy/arcticpy/trap_managers.py:143(restore)
494550/247275    1.871    0.000    1.966    0.000 arcticpy/trap_managers_utils.pyx:39(cy_n_trapped_electrons_from_watermarks)
487424/243712    1.910    0.000    1.942    0.000 arcticpy/trap_managers_utils.pyx:60(cy_update_watermark_volumes_for_cloud_below_highest)
670085/18615    0.736    0.000    1.778    0.000 /usr/lib/python3.8/copy.py:128(deepcopy)
   123996    1.030    0.000    1.728    0.000 /home/ash/Projects/arcticpy/arcticpy/trap_managers.py:522(<listcomp>)
   143925    0.085    0.000    1.722    0.000 /home/ash/Projects/arcticpy/arcticpy/trap_managers.py:252(fraction_of_traps_exposed_from_n_electrons)
55840/18614    0. . . .

```

I think `cy_clock_charge_in_one_direction` could maybe get faster and `collapse_redundant_watermarks`. But otherwise, there is nothing particularly jumping out at me. Suggestions and feedback very welcome!

**Comment by jkeger** [comment]
Hi Ash, thanks, this looks great! 

I'll have a proper play as soon as I can (likely not until next week) before feedback and probably pestering you with more questions :)
